### ISSUE TITLE
incorrectly escaped `%` when running sqlalchemy queries

### ISSUE THREAD

**Issue by amenck** [snowflakedb/snowflake-sqlalchemy#246]
Please answer these questions before submitting your issue. Thanks!

1. What version of Python are you using (`python --version`)?
`Python 3.7.11`

2. What operating system and processor architecture are you using (`python -c 'import platform; print(platform.platform())'`)? 
`Darwin-19.6.0-x86_64-i386-64bit`

3. What are the component versions in the environment (`pip list`)? 
[pip_list.txt](https://github.com/snowflakedb/snowflake-sqlalchemy/files/7355188/pip_list.txt)

4. What did you do?
I create the following table in my snowflake account:
```
CREATE OR REPLACE TABLE TEST.PUBLIC.TEST_DATE (DATE DATE);
INSERT INTO TEST.PUBLIC.TEST_DATE VALUES ('2020-01-01'), ('2020-02-01');
```
I then ran the following query on this table:
```
SELECT TO_CHAR(date, '%Y-%m-%d') AS DATE_STR FROM TEST.PUBLIC.TEST_DATE;
```
The output from the snowflake UI is exactly what I expected:
![Screen Shot 2021-10-15 at 1 53 04 PM](https://user-images.githubusercontent.com/11997462/137531423-d42c237f-a5c1-4e12-90d4-87325e3f45da.png)
I then ran the following python script:
```py
import sqlalchemy
from snowflake.sqlalchemy import URL as SNOWFLAKE_URL
from sqlalchemy.sql import text

sql = text(
    "select to_char(date, '%Y-%m-%d') as date "
    "from test.public.test_date"
)

url = SNOWFLAKE_URL(
    account="<redacted>",
    user="<redacted>",
    password="<redacted>",
)

engine = sqlalchemy.create_engine(url)

with engine.connect() as conn:
    result = conn.execute(sql)

print(list(result))
```

5. What did you expect to see?
The output of this script was:
```
[('%Y-%m-%d',), ('%Y-%m-%d',)]
```

6. What did you see instead?
I expected the script to return the same as I am seeing in the snowflake UI:
```
[('2020-01-01',), ('2020-02-01',)]
```

7. Can you set logging to DEBUG and collect the logs?
yep:
[snowflake_debug.log](https://github.com/snowflakedb/snowflake-sqlalchemy/files/7355210/snowflake_debug.log)

**Comment by sfc-gh-kterada** [comment]
@amenck, the Snowflake SQLAlchemy dialect doubles percent signs (%) by default like several other database dialects. To disable this behavior, please update your application as follows:

`...
with engine.connect() as conn:
    conn.dialect.identifier_preparer._double_percents = False
    result = conn.execute(sql)
...`

**Comment by amenck** [comment]
Hello @sfc-gh-kterada. The reason that SQLAlchemy doubles percent signs is to be able to deal with bound parameters in queries. I am not an expert on how the underlying base class works, but it is clear that setting `double_percents = False`, as you recommend, breaks queries that use bound parameters. Consider the following modification to the script above:

```
import sqlalchemy
from snowflake.sqlalchemy import URL as SNOWFLAKE_URL
from sqlalchemy.sql import text

sql = text(
    "select to_char(date, '%Y-%m-%d') as date "
    "from test.public.test_date "
    "where 1 = :bound_int"
).bindparams(bound_int=1)

url = SNOWFLAKE_URL(
    account="<redacted>",
    user="<redacted>",
    password="<redacted>",
)

engine = sqlalchemy.create_engine(url)

with engine.connect() as conn:
    conn.dialect.identifier_preparer._double_percents = False
    result = conn.execute(sql)

print(list(result))
```

This results in the query simply erroring out. Interestingly, removing the `conn.dialect.identifier_preparer._double_percents = False` fixes the issue and results in the correct output showing up: `[('2020-01-01',), ('2020-02-01',)]`.... so this behavior is really quite strange and inconsistent. Specifically, we see the following:

- in queries _with_ bound parameters, we must NOT set `double_percents = False` to get the correct output
- in queries _without_ bound parameters, we must set `double_percents = False` to get the correct output

Since we expect our code to work in all cases, I'm sure you can see that this workaround is insufficient. There seems to be a more fundamental bug with the underlying SQAlchemy adapter, and that must be root caused and fixed.

As a side note, it is worth noting that this same problem does *not* exist, for example, with the postgres SQLAlchemy adapter. I have tested `%` signs in queries for that adapter, and I do not see this incorrect behavior (specifically, I have tested `%` signs in `LIKE` statements, since postgres does not accept the `%Y` format). This problems seems to be specific to the Snowflake adapter.

**Comment by sfc-gh-mkeller** [comment]
After looking at https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#why-are-percent-signs-being-doubled-up-when-stringifying-sql-statements
I think the proper way to resolve this bug would be to do the following in your example @amenck:
```
...
with engine.connect() as conn:
    result = conn.execute(sql, (,))
```
In which case we bind no variables into the statement. This should then un-escape the double `%`s.

This would require the following changes: https://github.com/snowflakedb/snowflake-connector-python/pull/993

Does this sound like a good solution to you @amenck ?

**Comment by amenck** [comment]
@sfc-gh-mkeller yeah, that would work! If you make the fix to the python connector and release, I can adjust our code to use this workaround whenever there are no bound prams in the query object.

**Comment by sfc-gh-mkeller** [comment]
I have just merged the fix and our next release is planned for about 2 weeks from now.
We usually don't do hot-fix releases for bug fixes like this just in case the changes had some unforeseen consequences, so please hang tight until the next release. In the meantime we have a bunch of internal tests to be run until then.

We'll keep this issue open until you verify the final released fix!
